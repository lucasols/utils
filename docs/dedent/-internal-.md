[**@ls-stack/utils**](../README.md)

***

[@ls-stack/utils](../modules.md) / [dedent](README.md) / \<internal\>

# \<internal\>

## Type Aliases

### InterpolateValue

```ts
type InterpolateValue = string | number | boolean;
```

Defined in: [src/dedent.ts:3](https://github.com/lucasols/utils/blob/main/src/dedent.ts#L3)
