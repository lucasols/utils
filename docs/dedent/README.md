[**@ls-stack/utils**](../README.md)

***

[@ls-stack/utils](../modules.md) / dedent

# dedent

## Modules

- [\<internal\>](-internal-.md)

## Functions

### dedent()

#### Call Signature

```ts
function dedent(strings, ...values): string;
```

Defined in: [src/dedent.ts:5](https://github.com/lucasols/utils/blob/main/src/dedent.ts#L5)

##### Parameters

###### strings

`TemplateStringsArray`

###### values

...[`InterpolateValue`](-internal-.md#interpolatevalue)[]

##### Returns

`string`

#### Call Signature

```ts
function dedent(strings): string;
```

Defined in: [src/dedent.ts:9](https://github.com/lucasols/utils/blob/main/src/dedent.ts#L9)

##### Parameters

###### strings

`string`

##### Returns

`string`
