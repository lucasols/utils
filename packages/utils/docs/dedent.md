[**@ls-stack/utils**](README.md)

---

[@ls-stack/utils](modules.md) / dedent

# dedent

## Interfaces

### Dedent()

Defined in: [packages/utils/src/dedent.ts:6](https://github.com/lucasols/utils/blob/main/packages/utils/src/dedent.ts#L6)

#### Call Signature

```ts
Dedent(literals): string;
```

Defined in: [packages/utils/src/dedent.ts:7](https://github.com/lucasols/utils/blob/main/packages/utils/src/dedent.ts#L7)

##### Parameters

###### literals

`string`

##### Returns

`string`

#### Call Signature

```ts
Dedent(strings, ...values): string;
```

Defined in: [packages/utils/src/dedent.ts:8](https://github.com/lucasols/utils/blob/main/packages/utils/src/dedent.ts#L8)

##### Parameters

###### strings

`TemplateStringsArray`

###### values

...`unknown`[]

##### Returns

`string`

#### Properties

##### withOptions

```ts
withOptions: CreateDedent;
```

Defined in: [packages/utils/src/dedent.ts:9](https://github.com/lucasols/utils/blob/main/packages/utils/src/dedent.ts#L9)

---

### DedentOptions

Defined in: [packages/utils/src/dedent.ts:1](https://github.com/lucasols/utils/blob/main/packages/utils/src/dedent.ts#L1)

#### Properties

##### escapeSpecialCharacters?

```ts
optional escapeSpecialCharacters: boolean;
```

Defined in: [packages/utils/src/dedent.ts:2](https://github.com/lucasols/utils/blob/main/packages/utils/src/dedent.ts#L2)

##### trimWhitespace?

```ts
optional trimWhitespace: boolean;
```

Defined in: [packages/utils/src/dedent.ts:3](https://github.com/lucasols/utils/blob/main/packages/utils/src/dedent.ts#L3)

## Type Aliases

### CreateDedent()

```ts
type CreateDedent = (options) => Dedent;
```

Defined in: [packages/utils/src/dedent.ts:12](https://github.com/lucasols/utils/blob/main/packages/utils/src/dedent.ts#L12)

#### Parameters

##### options

[`DedentOptions`](#dedentoptions)

#### Returns

[`Dedent`](#dedent)

## Variables

### dedent

```ts
const dedent: Dedent;
```

Defined in: [packages/utils/src/dedent.ts:14](https://github.com/lucasols/utils/blob/main/packages/utils/src/dedent.ts#L14)
