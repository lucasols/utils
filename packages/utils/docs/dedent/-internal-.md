[**@ls-stack/utils**](../README.md)

***

[@ls-stack/utils](../modules.md) / [dedent](README.md) / \<internal\>

# \<internal\>

## Type Aliases

### InterpolationValue

```ts
type InterpolationValue = string | number | boolean | null | undefined;
```

Defined in: [packages/utils/src/dedent.ts:30](https://github.com/lucasols/utils/blob/main/packages/utils/src/dedent.ts#L30)
