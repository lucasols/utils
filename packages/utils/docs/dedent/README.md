[**@ls-stack/utils**](../README.md)

***

[@ls-stack/utils](../modules.md) / dedent

# dedent

## Modules

- [\<internal\>](-internal-.md)

## Interfaces

### Dedent()

Defined in: [packages/utils/src/dedent.ts:35](https://github.com/lucasols/utils/blob/main/packages/utils/src/dedent.ts#L35)

Dedent function interface that can be used both as a template tag and a regular function

#### Call Signature

```ts
Dedent(literals): string;
```

Defined in: [packages/utils/src/dedent.ts:39](https://github.com/lucasols/utils/blob/main/packages/utils/src/dedent.ts#L39)

Dedent function interface that can be used both as a template tag and a regular function

##### Parameters

###### literals

`string`

##### Returns

`string`

#### Call Signature

```ts
Dedent(strings, ...values): string;
```

Defined in: [packages/utils/src/dedent.ts:43](https://github.com/lucasols/utils/blob/main/packages/utils/src/dedent.ts#L43)

Dedent function interface that can be used both as a template tag and a regular function

##### Parameters

###### strings

`TemplateStringsArray`

###### values

...[`InterpolationValue`](-internal-.md#interpolationvalue)[]

##### Returns

`string`

#### Properties

##### withOptions

```ts
withOptions: CreateDedent;
```

Defined in: [packages/utils/src/dedent.ts:47](https://github.com/lucasols/utils/blob/main/packages/utils/src/dedent.ts#L47)

Create a new dedent function with custom options

***

### DedentOptions

Defined in: [packages/utils/src/dedent.ts:4](https://github.com/lucasols/utils/blob/main/packages/utils/src/dedent.ts#L4)

Configuration options for the dedent function behavior

#### Properties

##### escapeSpecialCharacters?

```ts
optional escapeSpecialCharacters: boolean;
```

Defined in: [packages/utils/src/dedent.ts:10](https://github.com/lucasols/utils/blob/main/packages/utils/src/dedent.ts#L10)

Whether to process escape sequences like \n, `, \$, and {
When true, allows using escaped characters in template literals

###### Default

```ts
true for template literals, false for plain strings
```

##### identInterpolations?

```ts
optional identInterpolations: boolean;
```

Defined in: [packages/utils/src/dedent.ts:21](https://github.com/lucasols/utils/blob/main/packages/utils/src/dedent.ts#L21)

Whether to preserve relative indentation of interpolated multi-line values
When true, multi-line interpolations are re-indented to match the surrounding context

###### Default

```ts
true
```

##### showNullishOrFalseValues?

```ts
optional showNullishOrFalseValues: boolean;
```

Defined in: [packages/utils/src/dedent.ts:27](https://github.com/lucasols/utils/blob/main/packages/utils/src/dedent.ts#L27)

Whether to display nullish or false values (false, null, undefined) in interpolations
When false, nullish or false values are skipped entirely

###### Default

```ts
false
```

##### trimWhitespace?

```ts
optional trimWhitespace: boolean;
```

Defined in: [packages/utils/src/dedent.ts:15](https://github.com/lucasols/utils/blob/main/packages/utils/src/dedent.ts#L15)

Whether to trim leading and trailing whitespace from the final result

###### Default

```ts
true
```

## Type Aliases

### CreateDedent()

```ts
type CreateDedent = (options) => Dedent;
```

Defined in: [packages/utils/src/dedent.ts:53](https://github.com/lucasols/utils/blob/main/packages/utils/src/dedent.ts#L53)

Factory function type for creating dedent functions with custom options

#### Parameters

##### options

[`DedentOptions`](#dedentoptions)

#### Returns

[`Dedent`](#dedent)

## Variables

### dedent

```ts
const dedent: Dedent;
```

Defined in: [packages/utils/src/dedent.ts:75](https://github.com/lucasols/utils/blob/main/packages/utils/src/dedent.ts#L75)

Remove common leading indentation from multi-line strings while preserving relative indentation.
Can be used as a tagged template literal or called with a plain string.

By default, it will dedent interpolated multi-line strings to match the surrounding context.
And it will not show falsy values.

#### Example

```typescript
const text = dedent`
  function hello() {
    console.log('world');
  }
`;
// Result:
"function hello() {
  console.log('world');
}"
```
