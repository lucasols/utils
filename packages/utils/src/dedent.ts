export interface DedentOptions {
  escapeSpecialCharacters?: boolean;
  trimWhitespace?: boolean;
}

export interface Dedent {
  (literals: string): string;
  (strings: TemplateStringsArray, ...values: unknown[]): string;
  withOptions: CreateDedent;
}

export type CreateDedent = (options: DedentOptions) => Dedent;

export const dedent: Dedent = createDedent({});

function createDedent(options: DedentOptions): Dedent {
  d.withOptions = (newOptions: DedentOptions): Dedent =>
    createDedent({ ...options, ...newOptions });

  return d;

  function d(literals: string): string;
  function d(strings: TemplateStringsArray, ...values: unknown[]): string;
  function d(strings: TemplateStringsArray | string, ...values: unknown[]) {
    const raw = typeof strings === 'string' ? [strings] : strings.raw;
    const {
      escapeSpecialCharacters = Array.isArray(strings),
      trimWhitespace = true,
    } = options;

    // first, perform interpolation
    let result = '';
    for (let i = 0; i < raw.length; i++) {
      let next = raw[i];

      if (escapeSpecialCharacters) {
        // handle escaped newlines, backticks, and interpolation characters
        next = next!
          .replace(/\\\n[ \t]*/g, '')
          .replace(/\\`/g, '`')
          .replace(/\\\$/g, '$')
          .replace(/\\\{/g, '{');
      }

      result += next;

      if (i < values.length) {
        // eslint-disable-next-line @typescript-eslint/restrict-plus-operands
        result += values[i];
      }
    }

    // now strip indentation
    const lines = result.split('\n');
    let mindent: null | number = null;
    for (const l of lines) {
      const m = l.match(/^(\s+)\S+/);
      if (m) {
        const indent = m[1]!.length;
        if (!mindent) {
          // this is the first indented line
          mindent = indent;
        } else {
          mindent = Math.min(mindent, indent);
        }
      }
    }

    if (mindent !== null) {
      const m = mindent; // appease TypeScript
      result = lines
        // https://github.com/typescript-eslint/typescript-eslint/issues/7140
        .map((l) => (l[0] === ' ' || l[0] === '\t' ? l.slice(m) : l))
        .join('\n');
    }

    // dedent eats leading and trailing whitespace too
    if (trimWhitespace) {
      result = result.trim();
    }

    // handle escaped newlines at the end to ensure they don't get stripped too
    if (escapeSpecialCharacters) {
      result = result.replace(/\\n/g, '\n');
    }

    return result;
  }
}
