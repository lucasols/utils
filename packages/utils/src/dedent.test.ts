import { describe, expect, test } from 'vitest';
import { dedent } from './dedent';

test('works without interpolation', () => {
  expect(
    dedent`
      first
      				 second
      				 third
    `,
  ).toMatchInlineSnapshot(`
    "first
    				 second
    				 third"
  `);
});

test('works with interpolation', () => {
  expect(
    dedent`
      first ${'line'}
      				 ${'second'}
      				 third
    `,
  ).toMatchInlineSnapshot(`
    "first line
    				 second
    				 third"
  `);
});

test('works with suppressed newlines', () => {
  expect(
    dedent`
      first \
      				 ${'second'}
      				 third
    `,
  ).toMatchInlineSnapshot(`
    "first second
    				 third"
  `);
});

test('works with blank first line', () => {
  expect(dedent`
    Some text that I might want to indent:
    	* reasons
    	* fun
    That's all.
  `).toMatchInlineSnapshot(`
    "Some text that I might want to indent:
    	* reasons
    	* fun
    That's all."
  `);
});

test('works with multiple blank first lines', () => {
  expect(
    dedent`

      first
      second
      third
    `,
  ).toMatchInlineSnapshot(`
    "first
    second
    third"
  `);
});

test('works with removing same number of spaces', () => {
  expect(
    dedent`
      first
      	second
      		 third
    `,
  ).toMatchInlineSnapshot(`
    "first
    	second
    		 third"
  `);
});

describe('single line input', () => {
  test('works with single line input', () => {
    expect(dedent`A single line of input.`).toMatchInlineSnapshot(
      `"A single line of input."`,
    );
  });

  test('works with single line and closing backtick on newline', () => {
    expect(dedent`
      A single line of input.
    `).toMatchInlineSnapshot(`"A single line of input."`);
  });

  test('works with single line and inline closing backtick', () => {
    expect(dedent`
      A single line of input.
    `).toMatchInlineSnapshot(`"A single line of input."`);
  });
});

test('can be used as a function', () => {
  expect(
    dedent(`
      A test argument.
    `),
  ).toMatchInlineSnapshot(`"A test argument."`);
});

describe('function character escapes', () => {
  describe('default behavior', () => {
    test('does not escape backticks', () => {
      expect(dedent(`\``)).toMatchInlineSnapshot(`"\`"`);
    });

    test('does not escape dollar signs', () => {
      expect(dedent(`$`)).toMatchInlineSnapshot(`"$"`);
    });

    test('does not escape opening braces', () => {
      expect(dedent(`{`)).toMatchInlineSnapshot(`"{"`);
    });

    test('escapes double-escaped backticks', () => {
      expect(dedent(`\\\``)).toMatchInlineSnapshot(`"\\\`"`);
    });

    test('escapes double-escaped dollar signs', () => {
      expect(dedent(`\\$`)).toMatchInlineSnapshot(`"\\$"`);
    });

    test('escapes double-escaped opening braces', () => {
      expect(dedent(`\\{`)).toMatchInlineSnapshot(`"\\{"`);
    });

    test('ignores closing braces', () => {
      expect(dedent(`}`)).toMatchInlineSnapshot(`"}"`);
    });
  });

  describe.each([undefined, false, true])(
    'with escapeSpecialCharacters %s',
    (escapeSpecialCharacters) => {
      test('backticks', () => {
        expect(
          dedent.withOptions({ escapeSpecialCharacters })(`\``),
        ).toMatchSnapshot();
      });

      test('dollar signs', () => {
        expect(
          dedent.withOptions({ escapeSpecialCharacters })(`$`),
        ).toMatchSnapshot();
      });

      test('opening braces', () => {
        expect(
          dedent.withOptions({ escapeSpecialCharacters })(`{`),
        ).toMatchSnapshot();
      });

      test('double-escaped backticks', () => {
        expect(
          dedent.withOptions({ escapeSpecialCharacters })(`\\\``),
        ).toMatchSnapshot();
      });

      test('double-escaped dollar signs', () => {
        expect(
          dedent.withOptions({ escapeSpecialCharacters })(`\\$`),
        ).toMatchSnapshot();
      });

      test('double-escaped opening braces', () => {
        expect(
          dedent.withOptions({ escapeSpecialCharacters })(`\\{`),
        ).toMatchSnapshot();
      });
    },
  );
});

describe.each([undefined, false, true])(
  'with trimWhitespace %s',
  (trimWhitespace) => {
    test('with trailing whitespace', () => {
      expect(
        dedent.withOptions({ trimWhitespace })(
          `
						foo---
						bar---
					`.replace(/-/g, ' '),
        ),
      ).toMatchSnapshot();
    });

    test('without trailing whitespace', () => {
      expect(
        dedent.withOptions({ trimWhitespace })(
          `
						foo
						bar
					`.replace(/-/g, ' '),
        ),
      ).toMatchSnapshot();
    });

    test('with leading whitespace', () => {
      expect(
        dedent.withOptions({ trimWhitespace })(`


						foo
					`),
      ).toMatchSnapshot();
    });
  },
);

describe('string tag character escapes', () => {
  describe('default behavior', () => {
    test('escapes backticks', () => {
      expect(dedent`\``).toMatchSnapshot();
    });

    test('escapes dollar signs', () => {
      expect(dedent`\$`).toMatchSnapshot();
    });

    test('escapes opening braces', () => {
      expect(dedent`\{`).toMatchSnapshot();
    });

    test('ignores closing braces', () => {
      expect(dedent`\}`).toMatchSnapshot();
    });
  });

  describe.each([undefined, false, true])(
    'with escapeSpecialCharacters %s',
    (escapeSpecialCharacters) => {
      test('backticks', () => {
        expect(
          dedent.withOptions({ escapeSpecialCharacters })`\``,
        ).toMatchSnapshot();
      });

      test('dollar signs', () => {
        expect(
          dedent.withOptions({ escapeSpecialCharacters })`\$`,
        ).toMatchSnapshot();
      });

      test('opening braces', () => {
        expect(
          dedent.withOptions({ escapeSpecialCharacters })`\{`,
        ).toMatchSnapshot();
      });
    },
  );
});

test("doesn't strip explicit newlines", () => {
  expect(dedent`
    <p>Hello world!</p>\n
  `).toMatchSnapshot();
});

test("doesn't strip explicit newlines with mindent", () => {
  expect(dedent`
    <p>
    	Hello world!
    </p>\n
  `).toMatchSnapshot();
});

test('works with spaces for indentation', () => {
  expect(
    dedent`
      first
        second
          third
    `,
  ).toMatchSnapshot();
});

test('works with tabs for indentation', () => {
  expect(
    dedent`
      first
      	second
      		third
    `,
  ).toMatchSnapshot();
});

test('works with escaped tabs for indentation', () => {
  expect(dedent('\t\tfirst\n\t\t\tsecond\n\t\t\t\tthird')).toMatchSnapshot();
});

test('does not replace \\n when called as a function', () => {
  expect(dedent(`\\nu`)).toBe('\\nu');
});

test('escapes escaped interpolation', () => {
  expect(dedent`\${'foo'}`).toMatchInlineSnapshot(`"\${'foo'}"`);
});

describe('recursive dedent (default)', () => {
  test('dedents interpolated multiline strings', () => {
    const nested = `nested line 1
nested line 2`;

    expect(dedent`
      outer line 1
        ${nested}
      outer line 2
    `).toMatchInlineSnapshot(`
      "outer line 1
        nested line 1
        nested line 2
      outer line 2"
    `);
  });

  test('preserves indentation of interpolated strings relative to insertion point', () => {
    const nested = `item 1
item 2`;

    expect(dedent`
      list:
        ${nested}
      end
    `).toMatchInlineSnapshot(`
      "list:
        item 1
        item 2
      end"
    `);
  });

  test('handles deeply nested interpolations', () => {
    const level3 = `level 3 content`;

    const level2 = dedent`
      level 2 start
        ${level3}
      level 2 end
    `;

    const result = dedent`
      level 1 start
        ${level2}
      level 1 end
    `;

    expect(result).toMatchInlineSnapshot(`
      "level 1 start
        level 2 start
          level 3 content
        level 2 end
      level 1 end"
    `);
  });

  test('handles single-line interpolations', () => {
    const singleLine = 'single line value';

    expect(dedent`
      before
        ${singleLine}
      after
    `).toMatchInlineSnapshot(`
      "before
        single line value
      after"
    `);
  });

  test('handles empty string interpolations', () => {
    const empty = '';

    expect(dedent`
      before
        ${empty}
      after
    `).toMatchInlineSnapshot(`
      "before
  
      after"
    `);
  });
});

const noIdentInterpolations = dedent.withOptions({
  identInterpolations: false,
});

describe('with identInterpolations false', () => {
  test('does not dedent interpolated multiline strings', () => {
    const nested = `nested line 1
nested line 2`;

    expect(noIdentInterpolations`
          outer line 1
            ${nested}
          outer line 2
    `).toMatchInlineSnapshot(`
      "outer line 1
        nested line 1
      nested line 2
      outer line 2"
    `);
  });
});

test('do not show falsey values', () => {
  expect(dedent`
    false:
    ${false}
    null:
    ${null}
    undefined:
    ${undefined}
    number:
    ${123}
    true:
    ${true}
    ---
  `).toMatchInlineSnapshot(`
    "false:

    null:

    undefined:

    number:
    123
    true:
    true
    ---"
  `);
});
