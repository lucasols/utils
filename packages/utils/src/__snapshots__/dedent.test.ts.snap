// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`doesn't strip explicit newlines 1`] = `
"<p>Hello world!</p>
"
`;

exports[`doesn't strip explicit newlines with mindent 1`] = `
"<p>
	Hello world!
</p>
"
`;

exports[`function character escapes > with escapeSpecialCharacters false > backticks 1`] = `"\`"`;

exports[`function character escapes > with escapeSpecialCharacters false > dollar signs 1`] = `"$"`;

exports[`function character escapes > with escapeSpecialCharacters false > double-escaped backticks 1`] = `"\\\`"`;

exports[`function character escapes > with escapeSpecialCharacters false > double-escaped dollar signs 1`] = `"\\$"`;

exports[`function character escapes > with escapeSpecialCharacters false > double-escaped opening braces 1`] = `"\\{"`;

exports[`function character escapes > with escapeSpecialCharacters false > opening braces 1`] = `"{"`;

exports[`function character escapes > with escapeSpecialCharacters true > backticks 1`] = `"\`"`;

exports[`function character escapes > with escapeSpecialCharacters true > dollar signs 1`] = `"$"`;

exports[`function character escapes > with escapeSpecialCharacters true > double-escaped backticks 1`] = `"\`"`;

exports[`function character escapes > with escapeSpecialCharacters true > double-escaped dollar signs 1`] = `"$"`;

exports[`function character escapes > with escapeSpecialCharacters true > double-escaped opening braces 1`] = `"{"`;

exports[`function character escapes > with escapeSpecialCharacters true > opening braces 1`] = `"{"`;

exports[`function character escapes > with escapeSpecialCharacters undefined > backticks 1`] = `"\`"`;

exports[`function character escapes > with escapeSpecialCharacters undefined > dollar signs 1`] = `"$"`;

exports[`function character escapes > with escapeSpecialCharacters undefined > double-escaped backticks 1`] = `"\\\`"`;

exports[`function character escapes > with escapeSpecialCharacters undefined > double-escaped dollar signs 1`] = `"\\$"`;

exports[`function character escapes > with escapeSpecialCharacters undefined > double-escaped opening braces 1`] = `"\\{"`;

exports[`function character escapes > with escapeSpecialCharacters undefined > opening braces 1`] = `"{"`;

exports[`string tag character escapes > default behavior > escapes backticks 1`] = `"\`"`;

exports[`string tag character escapes > default behavior > escapes dollar signs 1`] = `"$"`;

exports[`string tag character escapes > default behavior > escapes opening braces 1`] = `"{"`;

exports[`string tag character escapes > default behavior > ignores closing braces 1`] = `"\\}"`;

exports[`string tag character escapes > with escapeSpecialCharacters false > backticks 1`] = `"\\\`"`;

exports[`string tag character escapes > with escapeSpecialCharacters false > dollar signs 1`] = `"\\$"`;

exports[`string tag character escapes > with escapeSpecialCharacters false > opening braces 1`] = `"\\{"`;

exports[`string tag character escapes > with escapeSpecialCharacters true > backticks 1`] = `"\`"`;

exports[`string tag character escapes > with escapeSpecialCharacters true > dollar signs 1`] = `"$"`;

exports[`string tag character escapes > with escapeSpecialCharacters true > opening braces 1`] = `"{"`;

exports[`string tag character escapes > with escapeSpecialCharacters undefined > backticks 1`] = `"\`"`;

exports[`string tag character escapes > with escapeSpecialCharacters undefined > dollar signs 1`] = `"$"`;

exports[`string tag character escapes > with escapeSpecialCharacters undefined > opening braces 1`] = `"{"`;

exports[`with trimWhitespace false > with leading whitespace 1`] = `
"


foo
"
`;

exports[`with trimWhitespace false > with trailing whitespace 1`] = `
"
foo   
bar   
"
`;

exports[`with trimWhitespace false > without trailing whitespace 1`] = `
"
foo
bar
"
`;

exports[`with trimWhitespace true > with leading whitespace 1`] = `"foo"`;

exports[`with trimWhitespace true > with trailing whitespace 1`] = `
"foo   
bar"
`;

exports[`with trimWhitespace true > without trailing whitespace 1`] = `
"foo
bar"
`;

exports[`with trimWhitespace undefined > with leading whitespace 1`] = `"foo"`;

exports[`with trimWhitespace undefined > with trailing whitespace 1`] = `
"foo   
bar"
`;

exports[`with trimWhitespace undefined > without trailing whitespace 1`] = `
"foo
bar"
`;

exports[`works with escaped tabs for indentation 1`] = `
"first
	second
		third"
`;

exports[`works with spaces for indentation 1`] = `
"first
  second
    third"
`;

exports[`works with tabs for indentation 1`] = `
"first
	second
		third"
`;
